### ISSUE TITLE
Illegal Base64URL string

### ISSUE THREAD

**Issue by JMarkstrom** [MichaelGrafnetter/webauthn-interop#21]
Hi Michael,
Having installed v. `1.0.3` (passkey module only) and running command:

`Register-Passkey -UserId 'username@domain.com' -DisplayName 'YubiKey 5C NFC'`

Getting error: 

`ParentContainsErrorRecordException: Exception calling "Create" with "1" argument(s): "Illegal Base64URL string! (Parameter 'input')"`

In readme.md you state: 

> The [DSInternals.Passkeys](https://www.powershellgallery.com/packages/DSInternals.Passkeys) PowerShell module uses the DSInternals.Win32.WebAuthn library...

Should I understand this as I might be missing a dependency or that it is incorporated already in the PS module? 

**Environment**:
- Windows 11 24H2
- PowerShell 7.4.6
- YubiKey: any/all

Thank you in advance,

**Comment by aseigler** [comment]
That's this: https://x.com/alexseigler/status/1857127079705497882?t=lDqE2WGLCZavRrhX0qN5uQ&s=19

**Comment by JMarkstrom** [comment]
I just retested using the Yubico "PoC" scripts and I can successfully create, register and authenticate. 
But yes, I see there's a `0` at the end of the credential id now: 

`XDjx9wXDrVWBmE2MbPSkXdeMERfv6gN98uxavJ0ZrSyT8jYUs2BHKjAmh4P_T3Av0`

I will pass your Twitter post along.

**Comment by aseigler** [comment]
> I just retested using the Yubico "PoC" scripts and I can successfully create, register and authenticate. 
> But yes, I see there's a `0` at the end of the credential id now: 
> 
> `XDjx9wXDrVWBmE2MbPSkXdeMERfv6gN98uxavJ0ZrSyT8jYUs2BHKjAmh4P_T3Av0`
> 
> I will pass your Twitter post along.

It will work if you don't already have any authenticators registered on the account. The issue is with the `credentialId` being non-base64 in the `excludeCredentials`.

**Comment by JMarkstrom** [comment]
OK, I have passed it along. Note that my test user has multiple FIDO2 authenticators already however.

**Comment by aseigler** [comment]
> OK, I have passed it along. Note that my test user has multiple FIDO2 authenticators already however.

Yes, it kind of depends. Not all of the credentials are broken. The ones that are 44 chars are fine, the ones that are 64 chars got padded with a extra 0 for some reason so they are now 65 chars which is illegal (len % 1 is never OK).

**Comment by MichaelGrafnetter** [comment]
@aseigler Could you please send the bug report to fido-dev@microsoft.com ?

There are additional [confirmed issues](https://learn.microsoft.com/en-us/entra/identity/authentication/concept-fido2-compatibility#known-issues), including this one:
> If you registered more than three passkeys, sign in with a passkey might not work.

**Comment by aseigler** [comment]
Given that I haven't heard a peep from anyone on this I was thinking about this a little more and I looked at my data samples again.  It seems that all `credentialId` fields that used to be base64url with no padding now all either end in `0`, `1`, or `2`.  So, I am now suspicious that the number of padding characters required to turn the string into a legal base64 string has been encoded as that last character.

Like this

```C#
ReadOnlySpan<byte> input;
int paddingLength;

bool isFunky = BinaryPrimitives.TryReadInt32LittleEndian(input.Slice(input[^1], 1), out paddingLength);

if (!isFunky)
{
    paddingLength = (input.Length % 4) switch
    {
        0 => 0, // Padding is not needed
        2 => 2, // "==" missing in Base64Url vs. Base64
        3 => 1, // "=" missing in Base64Url vs. Base64
        _ => throw new ArgumentException("Illegal Base64URL string!", nameof(input))
    };
}
```

**Comment by aseigler** [comment]
Either that...or somebody was trying to do something like:

```C#
var b64 = "XDjx9wXDrVWBmE2MbPSkXdeMERfv6gN98uxavJ0ZrSyT8jYUs2BHKjAmh4P_T3Av";
int paddingChars = (b64.Length % 4);
```

And intended to do:
```C#
b64 = b64.PadRight(b64.Length + paddingChars, '=');
```

Which would yield `XDjx9wXDrVWBmE2MbPSkXdeMERfv6gN98uxavJ0ZrSyT8jYUs2BHKjAmh4P_T3Av`, but instead they did something like:

```C#
b64 = b64 + paddingChars;
```

Which would yield `XDjx9wXDrVWBmE2MbPSkXdeMERfv6gN98uxavJ0ZrSyT8jYUs2BHKjAmh4P_T3Av0` ...

**Comment by BatisteDONDOGLIO** [comment]
If I understand correctly, this error is thrown by the GraphAPI directly, and is therefore included as part of the API response, right? 

If not, could it be possible to parse the output ourselves and "fix" it accordingly, as a temporary workaround?

**Comment by aseigler** [comment]
We could certainly patch it temporarily. I was hesitant to do so because I assumed such a simple ridiculous error would have been corrected by now, but it has been broken for two months now.

I will put together a workaround.

**Comment by BatisteDONDOGLIO** [comment]
Indeed, I have been checking almost daily for MSGraph changelogs, hoping they would fix this quickly. But they still haven't fixed it. I'm not sure they even acknowledged this issue. 

In the meantime, a temporary workaround would be amazing, thank you @aseigler!

**Comment by grmp** [comment]
Not sure if anyone has mentioned this, but the padding only seems to be necessary if the user has already registered a passkey. It's still working for me if it's a "fresh" user.

**Comment by MichaelGrafnetter** [comment]
Hi @JMarkstrom and @BatisteDONDOGLIO  , the workaround implemented by @aseigler is out. We would really appreciate if you updated the PowerShell module to version 1.0.6 and told us if you still have any issues.

**Comment by cgaz90** [comment]
Hi @MichaelGrafnetter,

I successfully updated the DSInternals.Passkeys module to 1.0.6 and was able to provision security keys for users that previously encountered the "Illegal Base64URL string" error. The fix worked, but I faced an issue with the Update-Module cmdlet due to a mismatch in the Authenticode issuer.

`Install-Package: Authenticode issuer 'CN=Michael Grafnetter, O=Michael Grafnetter, L=Prague, C=CZ' of the new module 'DSInternals.Passkeys' with version '1.0.6' from root certificate authority 'CN=Microsoft Identity Verification Root Certificate Authority 2020, O=Microsoft Corporation, C=US' is not matching with the authenticode issuer 'CN=Michael Grafnetter, O=Michael Grafnetter, S="Praha, Hlavní město", C=CZ' of the previously-installed module 'DSInternals.Passkeys' with version '1.0.3' from root certificate authority 'CN=Sectigo Public Code Signing Root R46, O=Sectigo Limited, C=GB'.`

Workaround:

I had to uninstall the existing module and then reinstall it using:

```
Uninstall-Module -Name DSInternals.Passkeys
Install-Module -Name DSInternals.Passkeys -SkipPublisherCheck
```

**Comment by BatisteDONDOGLIO** [comment]
Hi @MichaelGrafnetter,

I can confirm the exact same issue @cgaz90 has mentioned and applied the same workaround. 
As for the core issue, I'll test it out in the coming days/weeks and report here. Thanks!

**Comment by MichaelGrafnetter** [comment]
> I successfully updated the DSInternals.Passkeys module to 1.0.6 and was able to provision security keys for users that previously encountered the "Illegal Base64URL string" error. The fix worked, but I faced an issue with the Update-Module cmdlet due to a mismatch in the Authenticode issuer.

Thanks @cgaz90 for confirmation. Yes, I switched from using long-term Sectigo certificates to short-lived ones issued by [Azure Trusted Signing](https://azure.microsoft.com/en-us/products/trusted-signing). I am a huge fan of this service, although, as it turns out, it has some issues. If NuGet and PowerShell execution policy compatibility is not improved soon, I will probably switch back.
